src/v/cluster/rm_stm.cc

```diff
@@ -1340,293 +1344,80 @@ ss::future<result<kafka_result>> rm_stm::idempotent_replicate(
-        }
-    }
-
-    // updating last observed seq, since we hold the mutex
-    // it's guranteed to be latest
-    session->tail_seq = bid.last_seq;
-
-    auto request = ss::make_lw_shared<inflight_request>();
-    request->last_seq = bid.last_seq;
-    request->is_processing = true;
-    session->cache.push_back(request);
-
-    // request comes in the right order, it's ok to replicate
-    std::optional<raft::replicate_stages> ss;
-    try {
-        ss = _raft->replicate_in_stages(synced_term, std::move(br), opts);
-        co_await std::move(ss->request_enqueued);
-    } catch (...) {
-        vlog(
-          _ctx_log.warn,
-          "[pid: {}] replication failed with {}",
-          bid.pid,
-          std::current_exception());
-    }
-
-    result<raft::replicate_result> r = errc::success;
-
-    if (!ss.has_value()) {
-        if (_raft->is_leader() && _raft->term() == synced_term) {
-            // we may not care about the requests waiting on the lock
-            // as soon as we release the lock the leader or term will
-            // be changed so the pending fibers won't pass the initial
-            // checks and be rejected with not_leader
-            co_await _raft->step_down("replicate_seq replication error");
-        }
-        u.return_all();
-        r = errc::replication_error;
-    } else {
-        try {
-            u.return_all();
-            enqueued->set_value();
-            r = co_await std::move(ss->replicate_finished);
-        } catch (...) {
-            vlog(
-              _ctx_log.warn,
-              "[pid: {}] replication failed with {}",
-              bid.pid,
-              std::current_exception());
-            r = errc::replication_error;
         }
-    }
-
-    // we don't need session->lock because we never interleave
-    // access to is_processing and offset with sync point (await)
-    request->is_processing = false;
-    if (r) {
-        auto old_offset = r.value().last_offset;
-        auto new_offset = from_log_offset(old_offset);
-        request->r = ret_t(kafka_result{new_offset});
-    } else {
-        request->r = ret_t(r.error());
-    }
-    for (auto& pending : request->parked) {
-        pending->set_value(request->r);
-    }
-    request->parked.clear();
-
-    if (!request->r) {
-        vlog(
-          _ctx_log.warn,
-          "[pid: {}] replication of batch with first seq: {} failed at "
-          "consensus level - error: {}",
-          bid.pid,
-          bid.first_seq,
-          request->r.error().message());
-        // if r was failed at the consensus level (not because has_failed)
-        // it should guarantee that all follow up replication requests fail
-        // too but just in case stepping down to minimize the risk
-        if (_raft->is_leader() && _raft->term() == synced_term) {
+        if (
+          _raft->is_leader()
+          && (result.error() == errc::replication_error || result.error() == errc::timeout)) {
```

Comment:
wrong condition for stepdown, any error must cause stepdown unless it already happen, check how it was done (is leader + term check)

Reply:
fixed.